deepspeed/runtime/bf16_optimizer.py

```diff
@@ -0,0 +1,582 @@
+                bf16_dp_partitions[partition_id].clone().float().detach())
+            self.fp32_groups_flat_partition[i].requires_grad = True
+
+            num_elem_list = [t.numel() for t in self.bf16_groups[i]]
+
+            # create fp32 gradients
+            self.fp32_groups_gradients_flat.append(
+                torch.zeros_like(self.bf16_groups_flat[i],
+                                 dtype=torch.float32))
+
+            # track individual fp32 gradients for entire model
+            fp32_gradients = self._split_flat_tensor(
+                flat_tensor=self.fp32_groups_gradients_flat[i],
+                num_elem_list=num_elem_list)
+            self.fp32_groups_gradients.append(fp32_gradients)
+
+            # flat tensor corresponding to actual fp32 gradients (i.e., minus alignment padding)
+            length_without_padding = sum(num_elem_list)
+            self.fp32_groups_actual_gradients_flat.append(
+                torch.narrow(self.fp32_groups_gradients_flat[i],
+                             0,
+                             0,
+                             length_without_padding))
+
+            # flat tensor corresponding to gradient partition
+            self.fp32_groups_gradient_flat_partition.append(
+                torch.narrow(self.fp32_groups_gradients_flat[i],
+                             0,
+                             partition_id * partition_size,
+                             partition_size))
+
+            # track fp32 gradient updates
+            self.fp32_groups_has_gradients.append([False] * len(self.bf16_groups[i]))
+
+            # Record padding required for alignment
+            if partition_id == dist.get_world_size(
+                    group=self.real_dp_process_group[i]) - 1:
+                padding = self.bf16_groups_flat[i].numel() - length_without_padding
+            else:
+                padding = 0
+
+            self.groups_padding.append(padding)
+
+            # update optimizer param groups to reference fp32 params partition
+            param_group['params'] = [self.fp32_groups_flat_partition[i]]
+
+            see_memory_usage(f'after initializing group {i}', force=True)
+
+        see_memory_usage('before initialize_optimizer', force=True)
+        self.initialize_optimizer_states()
+        see_memory_usage('end initialize_optimizer', force=True)
+
+        # Need optimizer states initialized before linking lp to optimizer state
+        self._link_all_hp_params()
+
+        see_memory_usage('end bf16_optimizer', force=True)
+
+    def _link_all_hp_params(self):
+        dp_world_size = dist.get_world_size(group=self.dp_process_group)
+        for i, param_group in enumerate(self.optimizer.param_groups):
+            # Link bf16 and fp32 params in partition
+            # TODO: Make this configurable
+            partition_id = dist.get_rank(group=self.real_dp_process_group[i])
+            partition_size = self.bf16_groups_flat[i].numel() // dp_world_size
```

Comment:
is `self.bf16_groups_flat[i].numel()` guaranteed to be a multiple of dp_world_size?

Reply:
Yes, because of [this](https://github.com/microsoft/DeepSpeed/blob/a9bfaee9c3b951f5a964ef69e82cfa984027fbb4/deepspeed/runtime/bf16_optimizer.py#L136-L139).